### PR TITLE
Add a method to TreePublisher to write json file and expose it in npg_tree_publisher script

### PR DESCRIPTION
**Pull request by marcomoscasgr**
This branch aims to add a method in the TreePublisher class. The function writes the irods top collection folder to json file only if the attribute mlwh_json is set. Issue #363.

### CODE REVIEW

File: lib/WTSI/NPG/HTS/TreePublisher.pm [hunk 1]
```diff
@@ -80,6 +91,10 @@ has 'require_checksum_cache' =>
                Function returning true for each file path to be published.
                CodeRef. Optional.
 
+               mlwh_json_cb
+               Callback writing information of a collection to JSON file.
```
Comment (kjsanger) [inline lib/WTSI/NPG/HTS/TreePublisher.pm:95]: "writing information about data objects to a JSON file"

File: lib/WTSI/NPG/HTS/TreePublisher.pm [hunk 2]
```diff
@@ -60,6 +64,13 @@ has 'require_checksum_cache' =>
      documentation => 'A list of file suffixes for which MD5 cache files ' .
                       'must be provided and will not be created on the fly');
 
+has 'mlwh_json' =>
```
Comment (kjsanger) [inline lib/WTSI/NPG/HTS/TreePublisher.pm:67]: This attribute should have a predicate "has_mlwh_json"

File: lib/WTSI/NPG/HTS/TreePublisher.pm [hunk 3]
```diff
@@ -103,9 +118,32 @@ has 'require_checksum_cache' =>
...
+      self->logconfess(q[could not close ml warehouse json file] .
+      qq[$self->mlwh_json]);
+    return 1;
+  }
+
   sub publish_tree {
     my ($self, $files) = $params->parse(@_);
 
+    if (defined $self->mlwh_json && defined $params->mlwh_json_cb) {
+      $self->logconfess('The mlwh_json_cb cannot be defined with mlwh_json variable set');
```
Comment (kjsanger) [inline lib/WTSI/NPG/HTS/TreePublisher.pm:145]: variable -> attribute

File: lib/WTSI/NPG/HTS/TreePublisher.pm [hunk 4]
```diff
@@ -171,6 +209,10 @@ has 'require_checksum_cache' =>
       $num_errors    += $ne;
     }
 
+    if (defined $self->mlwh_json) {
```
Comment (kjsanger) [inline lib/WTSI/NPG/HTS/TreePublisher.pm:212]: This should test the attribute's predicate.

**Review by kjsanger — CHANGES_REQUESTED** [review]
A few minor issues remaining to look at

File: lib/WTSI/NPG/HTS/TreePublisher.pm (hunk 2 above)
Comment (marcomoscasgr) [reply lib/WTSI/NPG/HTS/TreePublisher.pm:67]: Predicate "has_mlwh_json" added

File: lib/WTSI/NPG/HTS/TreePublisher.pm (hunk 1 above)
Comment (marcomoscasgr) [reply lib/WTSI/NPG/HTS/TreePublisher.pm]: Corrected

File: lib/WTSI/NPG/HTS/TreePublisher.pm [hunk 5]
```diff
@@ -80,6 +91,10 @@ has 'require_checksum_cache' =>
                Function returning true for each file path to be published.
                CodeRef. Optional.
 
+               mlwh_json_cb
+               Callback writing information of a collection to JSON file.
+               CodeRef. Optional. Cannot be used with mlwh_json attribute set.
```
[the remark this answers is not in the document]
Comment (marcomoscasgr) [reply lib/WTSI/NPG/HTS/TreePublisher.pm]: Corrected

File: lib/WTSI/NPG/HTS/TreePublisher.pm [hunk 6]
```diff
@@ -103,9 +118,32 @@ has 'require_checksum_cache' =>
...
+    close $json_fh or
+      self->logconfess(q[could not close ml warehouse json file] .
+      qq[$self->mlwh_json]);
+    return 1;
+  }
+
   sub publish_tree {
     my ($self, $files) = $params->parse(@_);
 
+    if (defined $self->mlwh_json && defined $params->mlwh_json_cb) {
```
[the remark this answers is not in the document]
Comment (marcomoscasgr) [reply lib/WTSI/NPG/HTS/TreePublisher.pm]: Replaced with predicate check

File: lib/WTSI/NPG/HTS/TreePublisher.pm (hunk 4 above)
Comment (marcomoscasgr) [reply lib/WTSI/NPG/HTS/TreePublisher.pm]: Replaced with predicate check

File: lib/WTSI/NPG/HTS/TreePublisher.pm (hunk 3 above)
Comment (marcomoscasgr) [reply lib/WTSI/NPG/HTS/TreePublisher.pm]: Corrected

**Review by kjsanger — APPROVED** [review]

**Pull request merged**